test/test_reductions.py

```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
+        t = make_tensor((10, 10), device, dtype)
+        self._test_noncontiguous(op, t[1::2, :-2:3])
+
+    @ops(reduction_ops)
+    def test_noncontiguous_transposed(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing a transposed tensor."""
+        t = make_tensor((10, 10), device, dtype)
+        self._test_noncontiguous(op, t.T)
+
+    @ops(reduction_ops)
+    def test_noncontiguous_expanded(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing a tensor with expanded singleton dimensions."""
+        t = make_tensor((10, 10), device, dtype)
+        self._test_noncontiguous(op, t.unsqueeze(1).expand(-1, 5, -1))
+
+    # NumPy does not support BFloat16 so we don't test that against reference
+    # implementations. We also don't compare dtypes or test for different
+    # keepdim because we already have other tests covering those.
+
+    def _test_ref(self, op: ReductionOpInfo, t: torch.Tensor, **reduction_kwargs):
+        """Compares op against op.ref for the given input and reduction kwargs"""
+        for args, kwargs in op.generate_args_kwargs(t, **reduction_kwargs):
+            kwargs.update(reduction_kwargs)
+            result = op(t, *args, **kwargs)
+            expected = op.ref(t.detach().cpu().numpy(), *args, **kwargs)
+            self.assertEqual(result, expected, exact_dtype=False)
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_scalar_input(self, device, dtype, op: ReductionOpInfo):
+        """Compares op against reference for scalar input tensors"""
+        self._test_ref(op, make_tensor([], device, dtype))
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_small_input(self, device, dtype, op: ReductionOpInfo):
+        """Compares op against reference for many small random input tensors"""
+        def make(*size):
+            return make_tensor(size, device, dtype)
+
+        self._test_ref(op, make(32))
+        self._test_ref(op, make(5, 2, 4, 3))
+        self._test_ref(op, make(5, 2, 4, 3), dim=0)
+        self._test_ref(op, make(5, 2, 4, 3), dim=-1)
+
+        if op.supports_multiple_dims:
+            self._test_ref(op, make(10, 1, 7, 4), dim=(0, -1))
+            self._test_ref(op, make(10, 1, 7, 4), dim=(0, 1, 3))
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_large_input(self, device, dtype, op: ReductionOpInfo):
+        """Compares op against reference for a very large input to check stability"""
+        self._test_ref(op, make_tensor((2 ** 20,), device, dtype))
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_extremal_values(self, device, dtype, op: ReductionOpInfo):
```

Comment:
Running this on integer types doesn't seem interesting. In the interest of test time float32 and complex64 is probably OK?